src/Picologic/AST.hs

```diff
@@ -98,10 +103,96 @@ cnf = simp . cnf' . nnf
 
 -- | Remove tautologies.
 simp :: Expr -> Expr
-simp ex = case ex of
-  -- Disj e1 (Neg e2) | e1 == e2 -> True
-  -- Disj (Neg e1) e2 | e1 == e2 -> True
-  Disj e1 e2 -> Disj (simp e1) (simp e2)
+simp = transformUp (propConst1 . simp1)
+
+simp1 :: Expr -> Expr
+simp1 ex = case ex of
+  Disj e1 (Neg e2) | e1 == e2 -> Top
+  Disj (Neg e1) e2 | e1 == e2 -> Top
   Conj e1 e2       | e1 == e2 -> e1
-                   | otherwise -> Conj (simp e1) (simp e2)
   e -> e
+
+-- | Test if expression is constant.
+isConst :: Expr -> Bool
+isConst Top = True
+isConst Bottom = True
+isConst e = False
+
+-- | Transform expression up from the bottom.
+transformUp :: (Expr -> Expr) -> Expr -> Expr
+-- TODO: This could probably be done with Data.Data, but it's outside my capabilities for now. 
+transformUp f ex = case ex of
+  Neg e -> f $ Neg (transformUp f e)
+  Conj e1 e2 -> f $ Conj (transformUp f e1) (transformUp f e2)
+  Disj e1 e2 -> f $ Disj (transformUp f e1) (transformUp f e2)
+  Iff e1 e2 -> f $ Iff (transformUp f e1) (transformUp f e2)
+  Implies e1 e2 -> f $ Implies (transformUp f e1) (transformUp f e2)
+  e -> f e
+
+-- | Transform expression down from the top.
+transformDown :: (Expr -> Expr) -> Expr -> Expr
+-- TODO: This could probably be done with Data.Data, but it's outside my capabilities for now. 
+transformDown f ex = case f ex of
+  Neg e -> Neg (transformDown f e)
+  Conj e1 e2 -> Conj (transformDown f e1) (transformDown f e2)
+  Disj e1 e2 -> Disj (transformDown f e1) (transformDown f e2)
+  Iff e1 e2 -> Iff (transformDown f e1) (transformDown f e2)
+  Implies e1 e2 -> Implies (transformDown f e1) (transformDown f e2)
+  e -> e
+
+-- | Convert expression to list of all subexpressions.
+toList :: Expr -> [Expr]
+-- TODO: This could probably be done with Data.Data, but it's outside my capabilities for now.
```

Comment:
You mean that remark about Data.Data? I am actually just curious how to do that.. Do you know how to do it? It seems there should be easy way to generate functions like that for general recursive data structures (ASTs), but I have no idea where to start.

Although I do agree this is perhaps more readable.